pgpy/pgp.py

```diff
@@ -1966,7 +1969,10 @@ def certify(self, subject, level=SignatureType.Generic_Cert, **prefs):
             if hash_prefs is not None:
                 sig._signature.subpackets.addnew('PreferredHashAlgorithms', hashed=True, flags=hash_prefs)
                 if sig.hash_algorithm is None:
-                    sig._signature.halg = hash_prefs[0]
+                    if len(hash_prefs):
```

Comment:
`if len(sequence)` is an anti-pattern in python because empty collections are falsy.

Probably what we should do here instead is change line 1969 from `if hash_prefs is not None:` to simply `if hash_prefs:`

Reply:
this doesn't quite solve the problem, though, because this stanza is doing two things:
 * it is adding a `PreferredHashAlgorithms` subpacket to the selfsig
 * it is selecting the hash algorithm for the selfsig itself

if we make the predicate for the whole stanza `if hash_prefs` then we have a case where `hash_prefs` is the empty list (or `None`) and `sig.hash_algorithm` is also `None`, which results in a signature being made without a chosen hash algorithm, right?

Reply:
I think i've resolved this second half by adding an additional fallback test outside of this one.